### ISSUE TITLE
Cannot compile ASN1 specs with ISO-8859 (Latin or Western) encoding

### ISSUE THREAD

**Issue by rmwesley** [pycrate-org/pycrate#17]
I am not asking for a fix. Just explaining an issue.

Here is the command I ran on bash:

```
$ ./tools/pycrate_asn1compile.py -i DSRC_instances_asn1_specs/EN15509/ -o DSRC_instances_asn1_specs/EN15509 -j
./tools/pycrate_asn1compile.py, args error: unable to read input file DSRC_instances_asn1_specs/EN15509/ISO14906Amd(2014)EfcDsrcGenericv5.asn
'utf-8' codec can't decode byte 0x93 in position 10503: invalid start byte
```
Sadly they surrounded the word UNIX time with left (“ = 0x93 in Latin1) and right (” = 0x94 in Latin1) double quotation marks instead of simple " in the `ISO14906Amd(2014)EfcDsrcGenericv5.asn` file in a comment, like so:
“UNIX time”

This makes up invalid UTF-8.
We find the same kind of issues in **EfcDsrcApplicationv5** and **AVIAEINumberingAndDataStructures** for souble quotation marks and other such characters, such as single quotation markas and dashes (’=0x94 and –=0x96).
**As recommended in the Wiki, comments should be manually removed.
I will simply change these characters by hand to their ASCII equivalents make up valid UTF-8 since they are just comments.**

It is impossible to detect 8-bit encodings programatically, right? Only if it is kept as metadata or noted down somewhere.
If the encoding could be determined, we could then simply do `open("myfile", encoding=determined_encoding)`.

I downloaded the ASN.1 specs directly from the ISO ASN.1 specs site.
The specifications using ISO-8859 encoding are https://standards.iso.org/iso/14906/ed-2/ISO14906Amd(2014)EfcDsrcGenericv5.asn and https://standards.iso.org/iso/14906/ed-2/ISO14906Amd(2014)EfcDsrcApplicationv5.asn.

**Comment by mitshell** [comment]
I agree that many ASN.1 specs provided here and there contains misencoded (or sometimes simply invalid) characters. This is generally the result of how the work is organized when building a technical standard or specification: different contributions from different companies and regions of the world are all merged in a big Word document, which then is eventually converted to PDF. This is error prone!
On the other side, the current compiler tries to decode any input as UTF-8 and breaks if it contains a non-UTF8 byte. We could potentially just drop those and keep decoding the rest of the file, and still get a working ASN.1 spec at the end.

**Comment by mitshell** [comment]
For my own recollection, this is happening here: https://github.com/pycrate-org/pycrate/blob/0b8309e258c59de9d0d7e6da31566097c92f224b/tools/pycrate_asn1compile.py#L200

**Comment by oopsbagel** [comment]
>     * convert wrongly encoded but meaningful characters to their expected UTF8 encoding.

There are inherent limitations that make it impossible to always correctly convert a wrongly encoded sequence, but there's a well crafted python library for fixing mojibake that works pretty well here: https://ftfy.readthedocs.io/en/latest/detect.html

Here's what this looks like minimally added to pycrate. I was able to decode and successfully compile @rmwesley 's example specifications.
```diff
diff --git a/tools/pycrate_asn1compile.py b/tools/pycrate_asn1compile.py
index d12b641..9c99e2e 100755
--- a/tools/pycrate_asn1compile.py
+++ b/tools/pycrate_asn1compile.py
@@ -33,6 +33,7 @@ import os
 import sys
 import argparse
 import inspect
+import ftfy
 
 from pycrate_asn1c.generator import _Generator
 from pycrate_asn1c.asnproc import (
@@ -97,6 +98,8 @@ def main():
                         help='provide an alternative python generator file path')
     parser.add_argument('-j', dest='json', action='store_true',
                         help='output a json file with information on ASN.1 objects dependency')
+    parser.add_argument('-G', action='store_true',
+                        help='guess the file encoding with python-ftfy')
     parser.add_argument('-fautotags', action='store_true',
                         help='force AUTOMATIC TAGS for all ASN.1 modules')
     parser.add_argument('-fextimpl', action='store_true',
@@ -196,15 +199,20 @@ def main():
         # read all file content into a single buffer
         txt = []
         for f in files:
+            open_file = lambda f: open(f, 'rb', encoding='utf-8')
+            read_file = lambda fd: fd.read()
+            if args.G:  # guess the encoding with ftfy
+                open_file = lambda f: open(f, 'rb')
+                read_file = lambda fd: ftfy.guess_bytes(fd.read())[0]
             try:
-                fd = open(f, 'r', encoding='utf-8')
+                fd = open_file(f)
             except Exception as e:
                 print('%s, args error: unable to open input file %s' % (sys.argv[0], f))
                 print(e)
                 return 1
             else:
                 try:
-                    txt.append( fd.read() )
+                    txt.append( read_file(fd) )
                 except Exception as e:
                     print('%s, args error: unable to read input file %s' % (sys.argv[0], f))
                     print(e)
```

I wouldn't want to include it exactly as is without better documentation around the feature, for reasons summed up in the documentation:
> Unlike the rest of ftfy, this may not be accurate, and it may create Unicode problems instead of solving them!

But it may be useful to people with poorly encoded asn1 files. Using ftfy on the command line is very simple: https://ftfy.readthedocs.io/en/latest/cli.html
```bash
$ ftfy -g bad.asn > good.asn
```
